### PR TITLE
compute-client: command module cleanup

### PR DESCRIPTION
**Pull request by teskje**
During moving the compute command protocol documentation into `mz_compute_client` I realized the the `command` module already contained a large amount of code, which would only grow as I added more comments. So I decided to clean it up a little first.

This PR does so by moving the `DataflowDescription` type and its contained types into a separate module `mz_compute_client::types`. This is in line with the storage client, which has a similar `types` module. This PR also moves the ID types `ReplicaId` and `ProcessId` into `mz_compute_client::controller`, where `ComputeInstanceId` already lives.

The end result is a `command` module that has 700 lines (roughly half of what it had before) and does not contain types used in the public API. It is now only imported by participants in the compute protocol (namely the compute controller and computed). We could make this module `pub(mz_compute)` if this was something Rust would support.

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

Sorry for the large diff! This is all just code movement, so hopefully easy to skim.

@vmarcos is the proper reviewer, with Lukas on vacation. @tokenrove is the random one.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Review by tokenrove — APPROVED** [review]

**Review by vmarcos — APPROVED** [review]
LGTM! I left a question on the protobuf declarations, though I have an inkling that it is not a blocker to move forward (just making sure to ask the question to double-check).

File: src/compute-client/src/types/dataflows.proto [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    repeated ProtoIndexExport index_exports = 4;
+    repeated ProtoSinkExport sink_exports = 5;
+    optional mz_repr.antichain.ProtoU64Antichain as_of = 6;
+    mz_repr.antichain.ProtoU64Antichain until = 7;
+    string debug_name = 8;
+}
+
+message ProtoIndexDesc {
+    mz_repr.global_id.ProtoGlobalId on_id = 1;
+    repeated mz_expr.scalar.ProtoMirScalarExpr key = 2;
```
[the remark this answers is not in the document]
Comment (teskje) [reply src/compute-client/src/types/dataflows.proto:59]: Yes, changing the numbers breaks backwards-compatibility. We don't usually care about that in our network protocols currently (see [the Backwards Compatibility Policy](https://www.notion.so/materialize/Backwards-compatibility-policy-434ecf894656435795ef3f61b3a54976#c839062780354d8b952fe22509168395)). 

You are right that there is some risk that we do durably store some of these types somewhere. I am not aware of us doing this for the `DataflowDescription` types and I have verified that a cluster with an index and an MV comes up successfully after updating from main to this branch. If that doesn't convince you, I'd be fine with not cleaning up the protobuf files and leaving them as before!

[the remark this answers is not in the document]
Comment (vmarcos) [reply src/compute-client/src/types/dataflows.proto:59]: Yes, this argument convinces me, and I think that manual testing is a good strategy to be extra-safe. Perhaps execute also an `EXPLAIN OPTIMIZER TRACE FOR MATERIALIZED VIEW <your-mv>` on your materialized view before and after to be 100% sure? I took a second look at the code and I am also quite confident that `DataflowDescription` is not written down anywhere. It is very heavily used by the explain / optimization machinery, though, so an extra (manual) test might be worth it.

[the remark this answers is not in the document]
Comment (teskje) [reply src/compute-client/src/types/dataflows.proto:59]: Depending on how the zero-downtime upgrades feature pans out, we might discover that we don't need backwards-compatibility for protocols between environmentd/computed/storaged at all. The "big version upgrades" approach discussed currently spawns a full new set of processes to run alongside, but separate from, the existing ones, which means processes of different versions never communicate. If we end up implementing this, we can think about getting rid of protobuf in favor or a simpler transport protocol.

**Pull request merged**